File: 백준/Gold/16933. 벽 부수고 이동하기 3/벽 부수고 이동하기 3.py
# https://www.acmicpc.net/problem/16933
import sys
from collections import deque
input = sys.stdin.readline
INF = 11


def bfs():
    go = ((-1, 0), (1, 0), (0, 1), (0, -1))
    q = deque()
    q.append((0, 0, 0, 1, 1))  # x y broken day cnt

    visited = [[[INF]*m for _ in range(n)] for _ in range(2)]
    visited[1][0][0] = 0

    while q:
        x, y, broken, day, cnt = q.popleft()
        if (x, y) == (n-1, m-1):
            return cnt

        if not day:
            for a, b in go:
                nx, ny = x+a, y+b
                if nx < 0 or nx >= n or ny < 0 or ny >= m:
                    continue
                if board[nx][ny]:
                    q.append((x, y, broken, 1, cnt+1))
                    break
                    
        for a, b in go:
            nx, ny = x+a, y+b
            if nx < 0 or nx >= n or ny < 0 or ny >= m:
                continue
            if broken >= visited[1-day][nx][ny]:
                continue

            if board[nx][ny]:
                if day:
                    if broken < k and broken+1 < visited[0][nx][ny]:
                        visited[0][nx][ny] = broken + 1
                        q.append((nx, ny, broken+1, 0, cnt+1))
            else:
                visited[1-day][nx][ny] = broken
                q.append((nx, ny, broken, 1-day, cnt+1))

    return -1


n, m, k = map(int, input().split())
board = [[int(x) for x in input().rstrip()] for _ in range(n)]

print(bfs())

File: 백준/Gold/16933. 벽 부수고 이동하기 3/벽 부수고 이동하기 3.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int[] nmk = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt)
                .toArray();

        int n = nmk[0];
        int m = nmk[1];
        int k = nmk[2];

        int[][] board = new int[n][];
        for (int i = 0; i < n; i++)
            board[i] = br.readLine().chars().map(num -> num-'0').toArray();

        System.out.println(bfs(board, n, m, k));
    }

    private static int bfs(int[][] board, int n, int m, int k) {
        int[] goX = {-1, 1, 0, 0};
        int[] goY = {0, 0, -1, 1};

        Queue<int[]> q = new ArrayDeque<>();
        q.add(new int[]{0, 0, 1, 0});

        boolean[][][] visited = new boolean[n][m][k+1];
        visited[0][0][0] = true;

        while (!q.isEmpty()) {
            int[] first = q.poll();
            int x = first[0];
            int y = first[1];
            int cnt = first[2];
            int broken = first[3];

            if (x == n-1 && y == m-1)
                return cnt;

            for (int goIdx = 0; goIdx < 4; goIdx++) {
                int nx = x + goX[goIdx];
                int ny = y + goY[goIdx];

                if (nx < 0 || nx >= n || ny < 0 || ny >= m)
                    continue;

                if (board[nx][ny] == 0) {
                    if (!visited[nx][ny][broken]) {
                        visited[nx][ny][broken] = true;
                        q.add(new int[]{nx, ny, cnt+1, broken});
                    }
                } else {
                    if (broken != k) {
                        if (cnt % 2 == 1) {
                            if (!visited[nx][ny][broken + 1]) {
                                visited[nx][ny][broken + 1] = true;
                                q.add(new int[]{nx, ny, cnt+1, broken + 1});
                            }
                        } else {
                            q.add(new int[]{x, y, cnt+1, broken});
                        }
                    }
                }
            }
        }

        return -1;
    }
}